src/compas/files/gltf.py

```
@@ -0,0 +1,629 @@
+
+        self._content = None
+        self._glb_buffer = None
+        self._buffers = {}
+
+        self.read()
+
+    def read(self):
+        with open(self.filepath, 'rb') as f:
+            self._content = memoryview(f.read())
+
+        is_glb = self._content[:4] == b'glTF'
+
+        if not is_glb:
+            content = self._content.tobytes().decode('utf-8')
+            self._content = None
+            self.json = json.loads(content)
+
+        else:
+            self.load_from_glb()
+            self._content = None
+
+        self.check_version()
+
+        if self.json:
+            for accessor in self.json.get('accessors', []):
+                accessor_data = self.access_data(accessor)
+                self.data.append(accessor_data)
+
+            for image in self.json.get('images', []):
+                image_data = self.get_image_data(image)
+                self.image_data.append(image_data)
+
+    def get_image_data(self, image):
+        if 'bufferView' not in image:
+            return None
+        buffer_view_index = image['bufferView']
+        buffer_view = self.json['buffersViews'][buffer_view_index]
+        buffer = self.get_buffer(buffer_view['buffer'])
+        offset = buffer_view.get('byteOffset', 0)
+        length_ = buffer_view['byteLength']
+        return buffer[offset: offset + length_]
+
+    def load_from_glb(self):
+        header = struct.unpack_from('<4sII', self._content)
+        file_size = header[2]
+
+        if file_size != len(self._content):
+            raise Exception('Bad glTF.  File size does not match.')
+
+        offset = 12
+
+        # load json
+        type_, length_, json_bytes, offset = self.load_chunk(offset)
+        if type_ != b'JSON':
+            raise Exception('Bad glTF.  First chunk not in JSON format')
+        json_str = json_bytes.tobytes().decode('utf-8')
+        self.json = json.loads(json_str)
+
+        # load binary buffer
+        if offset < len(self._content):
```

If that is not true, should we raise an exception or is it ok to just ignore?

---

This is checking whether there is a second chunk of binary data.  It may be that everything is stored as a data uri and so this chunk wouldn't be there.

---

👍